### PR TITLE
Hdist block voting

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

**Comment by almogdepaz** [comment]
@barakshani @zalmen this is without the abstain we can add it in a different pr

**Comment by antonlerner** [comment]
Please add unit tests to all new logic and refactored logic.

File: hare/hare.go [hunk 1]
```diff
@@ -213,19 +225,28 @@ var (
...
+	var results []types.BlockID
 	h.mu.RLock()
-	blks, ok := h.outputs[id]
-	if !ok {
-		h.mu.RUnlock()
-		return nil, ErrTooEarly
+	for id := lower; id <= upper; id++ {
+		blks, ok := h.outputs[id]
+		if !ok {
+			continue
```
Comment (antonlerner) [inline hare/hare.go:239]: log it?

File: miner/block_builder.go [hunk 2]
```diff
@@ -180,9 +182,15 @@ func (t *BlockBuilder) createBlock(id types.LayerID, atxID types.AtxId, eligibil
...
-		if err != nil {
-			return nil, errors.New(fmt.Sprintf("didn't receive hare result for layer %v %v", id-1, err))
+
+		bottom := types.LayerID(1)
+		if id > t.hdist {
+			bottom = id - t.hdist
+		}
+
+		votes, err = t.hareResult.GetResult(bottom, id-1)
+		if len(votes) == 0 {
```
Comment (antonlerner) [inline miner/block_builder.go]: votes can be nil, what is the expected behaviour then? I'd recommend checking err first

File: hare/hare.go (hunk 1 above)
Comment (almogdepaz) [reply hare/hare.go:239]: i think its redundant we have logs on stuck hare's and we have the results in the blocks
if we think theres an issue with the maAtuf and the buffer then maybe....

File: miner/block_builder.go (hunk 2 above)
Comment (almogdepaz) [reply miner/block_builder.go]: yes, thanks did that before i added the err

File: hare/hare.go [hunk 3]
```diff
@@ -213,19 +225,28 @@ var (
 )
 
 // GetResults returns the hare output for a given LayerID. returns error if we don't have results yet.
-func (h *Hare) GetResult(id types.LayerID) ([]types.BlockID, error) {
-	if h.isTooLate(InstanceId(id)) {
+func (h *Hare) GetResult(lower types.LayerID, upper types.LayerID) ([]types.BlockID, error) {
+
+	if h.isTooLate(InstanceId(lower)) {
```
Comment (zalmen) [inline hare/hare.go]: Why don't you just skip this one and return the rest? This is what you do below when you see missing results..

File: p2p/discovery/protocol.go [hunk 4]
```diff
@@ -37,7 +37,7 @@ const Name = "/udp/v2/discovery"
 const MessageBufSize = 100
 
 // MessageTimeout is the timeout we tolerate when waiting for a message reply
-const MessageTimeout = time.Second * 1 // TODO: Parametrize
+const MessageTimeout = time.Second * 2 // TODO: Parametrize
```
Comment (zalmen) [inline p2p/discovery/protocol.go:40]: Remind me, what stands behind this change?

File: hare/hare_test.go [hunk 5]
```diff
@@ -112,7 +112,7 @@ func TestHare_GetResult(t *testing.T) {
 
 	h := createHare(n1)
 
-	res, err := h.GetResult(types.LayerID(0))
+	res, err := h.GetResult(types.LayerID(0), types.LayerID(0))
```
Comment (gavraz) [inline hare/hare_test.go:115]: Please add new unit tests to the new GetResult func

File: hare/hare.go [hunk 6]
```diff
@@ -213,19 +225,28 @@ var (
 )
 
 // GetResults returns the hare output for a given LayerID. returns error if we don't have results yet.
-func (h *Hare) GetResult(id types.LayerID) ([]types.BlockID, error) {
-	if h.isTooLate(InstanceId(id)) {
+func (h *Hare) GetResult(lower types.LayerID, upper types.LayerID) ([]types.BlockID, error) {
```
Comment (gavraz) [inline hare/hare.go:228]: You never care which blocks belong to which layer? You just need the list of blocks right?

File: hare/hare.go [hunk 7]
```diff
@@ -112,13 +113,24 @@ func New(conf config.Config, p2p NetworkService, sign Signer, nid types.NodeId,
...
+	}
+	return false
+}
+
+func (h *Hare) oldestResultInBuffer() types.LayerID {
+
 	h.layerLock.RLock()
-	if int64(id) < int64(h.lastLayer)-int64(h.bufferSize) { // bufferSize>=0
+
+	if h.lastLayer <= types.LayerID(h.bufferSize) {
```
Comment (gavraz) [inline hare/hare.go:126]: Good fix.

File: hare/hare.go [hunk 8]
```diff
@@ -112,13 +113,24 @@ func New(conf config.Config, p2p NetworkService, sign Signer, nid types.NodeId,
...
 		h.layerLock.RUnlock()
-		return true
+		return 0
```
Comment (gavraz) [inline hare/hare.go:128]: Please add a const "firstLayer" or something like that.

**Review by gavraz — COMMENTED** [review]
I only reviewed the changes in the Hare. Looks good.

File: hare/hare.go [hunk 9]
```diff
@@ -112,13 +113,24 @@ func New(conf config.Config, p2p NetworkService, sign Signer, nid types.NodeId,
 }
 
 func (h *Hare) isTooLate(id InstanceId) bool {
+	if id < InstanceId(h.oldestResultInBuffer()) { // bufferSize>=0
+		return true
+	}
+	return false
+}
+
+func (h *Hare) oldestResultInBuffer() types.LayerID {
```
Comment (gavraz) [inline hare/hare.go:122]: add UT

File: miner/block_builder.go [hunk 10]
```diff
@@ -180,9 +182,15 @@ func (t *BlockBuilder) createBlock(id types.LayerID, atxID types.AtxId, eligibil
 	} else if id == config.Genesis+1 {
 		votes = append(votes, config.GenesisId)
 	} else {
-		votes, err = t.hareResult.GetResult(id - 1)
+
+		bottom := types.LayerID(1)
+		if id > t.hdist {
+			bottom = id - t.hdist
```
Comment (zalmen) [inline miner/block_builder.go]: The range of layers includes both the bottom and the id (as the higher limit). In such case this line should be `bottom = id -t.hdist + 1`

**Review by zalmen — CHANGES_REQUESTED** [review]
Is it well tested?

File: hare/hare_test.go (hunk 5 above)
Comment (almogdepaz) [reply hare/hare_test.go:115]: i changed the original one

File: p2p/discovery/protocol.go (hunk 4 above)
Comment (almogdepaz) [reply p2p/discovery/protocol.go:40]: discovery tests failed on travis, talked to yosher and he suggested this

Comment (zalmen) [reply p2p/discovery/protocol.go:40]: @y0sher, is it a temporary fix or can it be pushed to develop this way?

**Review by zalmen — APPROVED** [review]

**Pull request merged**